pkg/repository/maintenance/maintenance.go

```diff
@@ -0,0 +1,494 @@
+				repoJobConfigKey)
+		}
+	}
+
+	if _, ok := cm.Data[GlobalKeyForRepoMaintenanceJobCM]; ok {
+		logger.Debugf("Find the global repo maintenance config for repo %s", repo.Name)
+
+		if result == nil {
+			result = new(JobConfigs)
+		}
+
+		globalResult := new(JobConfigs)
+
+		if err := json.Unmarshal([]byte(cm.Data[GlobalKeyForRepoMaintenanceJobCM]), globalResult); err != nil {
+			return nil, errors.Wrapf(
+				err,
+				"fail to unmarshal configs from %s's key %s",
+				repoMaintenanceJobConfig,
+				GlobalKeyForRepoMaintenanceJobCM)
+		}
+
+		if result.PodResources == nil && globalResult.PodResources != nil {
+			result.PodResources = globalResult.PodResources
+		}
+
+		if len(result.LoadAffinities) == 0 {
+			result.LoadAffinities = globalResult.LoadAffinities
+		}
+	}
+
+	return result, nil
+}
+
+// WaitJobComplete waits the completion of the specified maintenance job and return the BackupRepositoryMaintenanceStatus
+func WaitJobComplete(cli client.Client, ctx context.Context, jobName, ns string, logger logrus.FieldLogger) (velerov1api.BackupRepositoryMaintenanceStatus, error) {
+	log := logger.WithField("job name", jobName)
+
+	maintenanceJob, err := waitForJobComplete(ctx, cli, ns, jobName)
+	if err != nil {
+		return velerov1api.BackupRepositoryMaintenanceStatus{}, errors.Wrap(err, "error to wait for maintenance job complete")
+	}
+
+	log.Info("Maintenance repo complete")
+
+	result, err := getResultFromJob(cli, maintenanceJob)
+	if err != nil {
+		log.WithError(err).Warn("Failed to get maintenance job result")
+	}
+
+	return composeStatusFromJob(maintenanceJob, result), nil
+}
+
+// WaitAllJobsComplete checks all the incomplete maintenance jobs of the specified repo and wait for them to complete,
+// and then return the maintenance jobs' status in the range of limit
+func WaitAllJobsComplete(ctx context.Context, cli client.Client, repo *velerov1api.BackupRepository, limit int, log logrus.FieldLogger) ([]velerov1api.BackupRepositoryMaintenanceStatus, error) {
+	jobList := &batchv1.JobList{}
+	err := cli.List(context.TODO(), jobList, &client.ListOptions{
+		Namespace: repo.Namespace,
+	},
+		client.MatchingLabels(map[string]string{RepositoryNameLabel: repo.Name}),
+	)
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "error listing maintenance job for repo %s", repo.Name)
+	}
+
+	if len(jobList.Items) == 0 {
+		return nil, nil
+	}
+
+	sort.Slice(jobList.Items, func(i, j int) bool {
+		return jobList.Items[i].CreationTimestamp.Time.Before(jobList.Items[j].CreationTimestamp.Time)
+	})
+
+	history := []velerov1api.BackupRepositoryMaintenanceStatus{}
+
+	startPos := len(history) - limit
+	if startPos < 0 {
+		startPos = 0
+	}
+
+	for i := startPos; i < len(jobList.Items); i++ {
+		job := &jobList.Items[i]
+
+		if job.Status.Succeeded == 0 && job.Status.Failed == 0 {
+			log.Infof("Waiting for maintenance job %s to complete", job.Name)
+
+			updated, err := waitForJobComplete(ctx, cli, job.Namespace, job.Name)
```

Comment:
Is there a timeout for this "wait"?  
The concern is that if a job's status is not updated for some reason we'll never have a chance to run another maintenance job.

Reply:
As the discussion, we will print some logs periodically, so that users know this status when there is no maintenance job for a long time and take action to solve the problem.

Reply:
Added the log mechanism